src/scenic/core/object_types.py

```diff
@@ -3,666 +3,1760 @@
+
+    @cached_property
+    def surface(self):
+        """A region containing the entire surface of this object"""
+        return self.occupiedSpace.getSurfaceRegion()
+
+    @cached_property
+    def onSurface(self):
+        """The surface used by the ``on`` specifier.
+
+        This region is used to sample position when
+        another object is placed ``on`` this object. By default
+        the top surface of this object (`topSurface`), but can
+        be overwritten by subclasses.
+        """
+        return self.topSurface
+
+    @cached_property
+    def rightSurface(self):
+        """A region containing the right surface of this object"""
+        return defaultSideSurface(self.occupiedSpace, dimension=0, \
+            positive=True, thresholds=self.sideComponentThresholds)
+
+    @cached_property
+    def leftSurface(self):
+        """A region containing the left surface of this object"""
+        return defaultSideSurface(self.occupiedSpace, dimension=0, \
+            positive=False, thresholds=self.sideComponentThresholds)
+
+    @cached_property
+    def frontSurface(self):
+        """A region containing the front surface of this object"""
+        return defaultSideSurface(self.occupiedSpace, dimension=1, \
+            positive=True, thresholds=self.sideComponentThresholds)
+
+    @cached_property
+    def backSurface(self):
+        """A region containing the back surface of this object"""
+        return defaultSideSurface(self.occupiedSpace, dimension=1, \
+            positive=False, thresholds=self.sideComponentThresholds)
+
+    @cached_property
+    def topSurface(self):
+        """A region containing the top surface of this object"""
+        return defaultSideSurface(self.occupiedSpace, dimension=2, \
+            positive=True, thresholds=self.sideComponentThresholds)
+
+    @cached_property
+    def bottomSurface(self):
+        """A region containing the bottom surface of this object"""
+        return defaultSideSurface(self.occupiedSpace, dimension=2, \
+            positive=False, thresholds=self.sideComponentThresholds)
+
+    def show_3d(self, viewer, highlight=False):
+        if needsSampling(self):
+            raise RuntimeError('tried to show() symbolic Object')
+
+        # Render the object
+        object_mesh = self.occupiedSpace.mesh.copy()
+
+        if highlight:
+            object_mesh.visual.face_colors = [30, 179, 0, 255]
+        elif self.color is not None:
+            object_mesh.visual.face_colors = self.color
+
+        viewer.add_geometry(object_mesh)
+
+        # If the camera is not a sphere, render the visible pyramid as a blue wireframe
+        if self.viewAngles != (math.tau, math.pi) or self.visibleDistance != 50:
+            camera_pyramid_mesh = self.visibleRegion.mesh.copy()
+
+            edges = camera_pyramid_mesh.face_adjacency_edges[camera_pyramid_mesh.face_adjacency_angles > np.radians(0.1)].copy()
+            vertices = camera_pyramid_mesh.vertices.copy()
```

Comment:
Are all these copies necessary? It doesn't look like you're doing any mutations. If we do need them it would be good to have an explanatory comment.

Reply:
Definitely not all necessary, I removed all but one.

Reply:
Why do we even need one?